code/game/machinery/kitchen/microwave.dm

```diff
@@ -153,237 +161,250 @@
+			highest_count = recipe.complexity
+			. = recipe
 
 /obj/machinery/microwave/proc/cook()
+	cook_break = FALSE
+	cook_dirty = FALSE
+
 	if(stat & (NOPOWER|BROKEN))
 		return
-	start()
-	if (reagents.total_volume==0 && !contents.len) //dry run
-		if (!wzhzhzh(10))
-			abort()
-			return
-		stop()
+
+	if (!reagents.total_volume && !length(get_contained_external_atoms())) //dry run
+		start()
 		return
 
-	var/decl/recipe/recipe = select_recipe()
-	var/obj/cooked
+	if (reagents.total_volume && prob(50)) // 50% chance a liquid recipe gets messy
+		dirty += Ceiling(reagents.total_volume / 10)
+
+	var/decl/recipe/recipe = src.select_recipe()
 	if (!recipe)
-		dirty += 1
-		if (prob(max(10,dirty*5)))
-			if (!wzhzhzh(4))
-				abort()
-				return
-			muck_start()
-			wzhzhzh(4)
-			muck_finish()
-			cooked = fail()
-			cooked.dropInto(loc)
-			return
+		failed = TRUE
+		cook_time = update_cook_time()
+		dirty += 5
+		if (prob(max(10, dirty*5)))
+			// It's dirty enough to mess up the microwave
+			cook_dirty = TRUE
 		else if (has_extra_item())
-			if (!wzhzhzh(4))
-				abort()
-				return
-			broke()
-			cooked = fail()
-			cooked.dropInto(loc)
-			return
-		else
-			if (!wzhzhzh(10))
-				abort()
-				return
-			stop()
-			cooked = fail()
-			cooked.dropInto(loc)
-			return
+			// Something's in the microwave that shouldn't be! Time to break!
+			cook_break = TRUE
 	else
-		var/halftime = round(recipe.time/10/2)
-		if (!wzhzhzh(halftime))
-			abort()
-			return
-		if (!wzhzhzh(halftime))
-			abort()
-			cooked = fail()
-			cooked.dropInto(loc)
-			return
-		cooked = recipe.make_food(src)
-		LAZYCLEARLIST(ingredients)
+		failed = FALSE
+		cook_time = update_cook_time(round(recipe.time * 2))
+
+	start()
+
+/obj/machinery/microwave/proc/update_cook_time(var/ct = 200)
+	return (ct / cooking_power)
+
+/obj/machinery/microwave/proc/finish_cooking()
+	var/decl/recipe/recipe = src.select_recipe()
+	if(!recipe)
+		return
+	var/result = recipe.result
+	var/list/cooked_items = list()
+	while(recipe)
+		cooked_items += recipe.make_food(src)
+		recipe = src.select_recipe()
+		if (!recipe || (recipe.result != result))
+			break
+
+	//Any leftover reagents are divided amongst the foods
+	var/total = reagents.total_volume
+	for (var/obj/item/chems/food/snacks/S in cooked_items)
+		reagents.trans_to_holder(S.reagents, total/cooked_items.len)
+		S.dropInto(loc) // since eject only ejects ingredients!
+
+	dispose(message = FALSE) //clear out anything left
+
+	return
+
+/obj/machinery/microwave/Process() // What you see here are the remains of proc/wzhzhzh, 2011 - 2021. RIP.
+	if (!operating || stat & (NOPOWER|BROKEN))
 		stop()
-		if(cooked)
-			cooked.dropInto(loc)
 		return
 
-// Behold: the worst proc name in the codebase.
-/obj/machinery/microwave/proc/wzhzhzh(var/seconds)
-	for (var/i=1 to seconds)
-		if (stat & (NOPOWER|BROKEN))
-			return 0
-		use_power_oneoff(500)
-		sleep(10)
-	return 1
+	use_power_oneoff(active_power_usage)
```

Comment:
The point of this var is that if you set the power to active, it will _automatically_ do this every tick (in a more performant manner).